File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int counter = 0;
        ListNode temp = head;
        
        while (true){
            if (temp == null){
                break;
            }
            counter++;
            temp = temp.next;
        }
        
        if (counter == 1){
            return null;
        }
        else if (counter == n){
            return head = head.next;
        }
        ListNode pre = head;
        temp = pre.next;
        int i = 1;
        while (true){
            if (counter - n == i){
                pre.next = temp.next;
                break;
            }
            if (temp == null){
                pre.next = null;
                break;
            }
            i++;
            pre = pre.next;
            temp = temp.next;
        }
        return head;
    }
}

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = 0;
        
        ListNode* cur = head;
        
        while(cur != NULL){
            len++;
            cur = cur->next;
        }
        
        cur = head;
        ListNode* prev = NULL;
        n = len - n;
        len = 0;
        while(cur != NULL){
            if(len == n){
                if(prev == NULL){
                    return head->next;
                }
                else{
                    prev->next = cur->next;
                    return head;
                }
            }
            prev = cur;
            cur = cur->next;
            len++;
        }
        return head;
    }
};